presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalMergeSourceOperator.java

```diff
@@ -0,0 +1,180 @@
+import com.facebook.presto.operator.DriverContext;
+import com.facebook.presto.operator.Operator;
+import com.facebook.presto.operator.OperatorContext;
+import com.facebook.presto.operator.OperatorFactory;
+import com.facebook.presto.operator.PageWithPositionComparator;
+import com.facebook.presto.operator.PageWithPositionComparatorFactory;
+import com.facebook.presto.operator.WorkProcessor;
+import com.facebook.presto.operator.exchange.LocalExchange.LocalExchangeFactory;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.SortOrder;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.util.MergeSortedPages.mergeSortedPages;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class LocalMergeSourceOperator
+        implements Operator
+{
+    public static class LocalMergeSourceOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final PageWithPositionComparatorFactory comparatorFactory;
+        private final LocalExchangeFactory localExchangeFactory;
+        private final List<Type> types;
+        private final List<Integer> sortChannels;
+        private final List<SortOrder> orderings;
+        private boolean closed;
+
+        public LocalMergeSourceOperatorFactory(
+                int operatorId,
+                PlanNodeId planNodeId,
+                PageWithPositionComparatorFactory comparatorFactory,
+                LocalExchangeFactory localExchangeFactory,
+                List<Type> types,
+                List<Integer> sortChannels,
+                List<SortOrder> orderings)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.comparatorFactory = requireNonNull(comparatorFactory, "comparator is null");
+            this.localExchangeFactory = requireNonNull(localExchangeFactory, "exchange is null");
+            this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+            this.sortChannels = ImmutableList.copyOf(requireNonNull(sortChannels, "sortChannels is null"));
+            this.orderings = ImmutableList.copyOf(requireNonNull(orderings, "orderings is null"));
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+
+            LocalExchange inMemoryExchange = localExchangeFactory.getLocalExchange(driverContext.getLifespan());
```

Comment:
nit: why `inMemoryExchange` and not just `localExchange`? Is is not the case that every local exchange is an in-memory exchange?